File: python/82_path_sum_three_ways.py
def path_sum_three_ways(matrix: list) -> int:
    r = len(matrix)
    c = len(matrix[0])
    min_sums = [matrix[row][0] for row in range(r)]
    
    for col in range(1, c):
        for row in range(0, r):
            # calculate path sum after moving right
            min_sums[row] += matrix[row][col] 
            
            if row > 0:
                # compare with path sum after moving down
                min_sums[row] = min(min_sums[row], min_sums[row - 1] + matrix[row][col])
        
        for row in range(r - 2, -1, -1):
            # compare with path sum after moving up
            min_sums[row] = min(min_sums[row], min_sums[row + 1] + matrix[row][col])
    
    return min(min_sums)
    

if __name__ == "__main__":
    f = open("82_matrix.txt", 'r')
    matrix = []
    for line in f.readlines():
        matrix.append([int(num) for num in line.strip().split(",")])
    f.close()
    
    # matrix = [
        # [131, 673, 234, 103, 18],
        # [201, 96, 342, 965, 150],
        # [630, 803, 746, 422, 111],
        # [537, 699, 497, 121, 956],
        # [805, 732, 524, 37, 331]
    # ]
    
    print(path_sum_three_ways(matrix))

File: cpp/82_path_sum_three_ways.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <cstring>

int path_sum_three_ways(std::vector<std::vector<int>> matrix){
    int rows = matrix.size(), cols = matrix[0].size();
    
    // make copy of first column of matrix
    std::vector<int> min_sums;
    for (int r = 0; r < rows; r++){
        min_sums.push_back(matrix[r][0]);
    }
    
    for (int c = 0; c < cols; c++){
        for (int r = 0; r < rows; r++){
            // minimum path sum after moving right from left neighbor
            min_sums[r] += matrix[r][c];
            
            // check if approaching matrix[r][c] from above has smaller path sum
            if (r > 0){
                min_sums[r] = std::min(min_sums[r], min_sums[r - 1] + matrix[r][c]);
            }
        }
        
        // check if approaching matrix[r][c] from below has smaller path sum
        for (int r = rows - 2; r >= 0; r--){
            min_sums[r] = std::min(min_sums[r], min_sums[r + 1] + matrix[r][c]);
        }
    }
    int min_path = min_sums[0];
    for (int sum : min_sums){
        min_path = std::min(min_path, sum);
    }
    return min_path;
}

int main(){
    std::ifstream file("/uploads/82_matrix.txt");
    std::string s;
    std::vector<std::vector<int>> matrix;
    
    // tokenize each line into vector of ints
    while (std::getline(file, s)){
        std::vector<int> row;
        char *pch = std::strtok(&s[0], ",");
        while (pch != NULL){
            std::string num(pch);
            row.push_back(stoi(num));
            pch = std::strtok(NULL, ",");
        }
        matrix.push_back(row);
    }
    file.close();
    
    std::cout << path_sum_three_ways(matrix) << "\n";
}
